integration/querier_remote_read_test.go

```diff
@@ -131,118 +135,228 @@ func TestQuerierRemoteRead(t *testing.T) {
+				Step:  1,
+				Start: startMs,
+				End:   endMs,
+			})
+			require.NoError(t, err)
+
+			req := &prompb.ReadRequest{
+				Queries:               []*prompb.Query{q},
+				AcceptedResponseTypes: []prompb.ReadRequest_ResponseType{prompb.ReadRequest_STREAMED_XOR_CHUNKS},
+			}
+
+			data, err := proto.Marshal(req)
+			require.NoError(t, err)
+			compressed := snappy.Encode(nil, data)
+
+			// Call the remote read API endpoint with a timeout.
+			httpReqCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+
+			httpReq, err := http.NewRequestWithContext(httpReqCtx, "POST", "http://"+querier.HTTPEndpoint()+"/prometheus/api/v1/read", bytes.NewReader(compressed))
+			require.NoError(t, err)
+			httpReq.Header.Add("Accept-Encoding", "snappy")
+			httpReq.Header.Set("X-Scope-OrgID", "user-1")
+			httpReq.Header.Set("X-Prometheus-Remote-Read-Version", "0.1.0")
+
+			httpResp, err := http.DefaultClient.Do(httpReq)
+			require.NoError(t, err)
+			require.Equal(t, http.StatusOK, httpResp.StatusCode)
+
+			// Fetch streaming response
+			stream := remote.NewChunkedReader(httpResp.Body, remote.DefaultChunkedReadLimit, nil)
+
+			var results []prompb.ChunkedReadResponse
+			for {
+				var res prompb.ChunkedReadResponse
+				err := stream.NextProto(&res)
+				if errors.Is(err, io.EOF) {
+					break
+				}
+				require.NoError(t, err)
+				results = append(results, res)
+			}
+
+			// Validate the returned remote read data
+			sampleIdx := 0
+			for _, result := range results {
+				// We're only expected a single series `series_1`.
+				require.Len(t, result.ChunkedSeries, 1)
+				require.Equal(t, "series_1", result.ChunkedSeries[0].Labels[0].GetValue())
+
+				for _, rawChk := range result.ChunkedSeries[0].Chunks {
+					var enc chunkenc.Encoding
+					switch rawChk.Type {
+					case prompb.Chunk_XOR:
+						enc = chunkenc.EncXOR
+					case prompb.Chunk_HISTOGRAM:
+						enc = chunkenc.EncHistogram
+					// TODO: Once prompb.Chunk_FLOAT_HISTOGRAM gets added in https://github.com/prometheus/prometheus/pull/12085 we can switch to using that
+					case prompb.Chunk_Encoding(3):
+						enc = chunkenc.EncFloatHistogram
+					default:
+						require.Fail(t, "unrecognized chunk type")
+					}
+
+					chk, err := chunkenc.FromData(enc, rawChk.Data)
+					require.NoError(t, err)
+
+					chkItr := chk.Iterator(nil)
+					chkIdx := 0
+					for valType := chkItr.Next(); valType != chunkenc.ValNone; valType = chkItr.Next() {
+						require.Equal(t, tc.expectedValType, valType)
+						switch valType {
+						case chunkenc.ValFloat:
+							ts, val := chkItr.At()
+							require.Equal(t, samples[sampleIdx].Timestamp, ts)
+							require.Equal(t, samples[sampleIdx].Value, val)
+						case chunkenc.ValHistogram:
+							ts, h := chkItr.AtHistogram()
+							require.Equal(t, histograms[sampleIdx].Timestamp, ts)
+
+							expected := remote.HistogramProtoToHistogram(histograms[sampleIdx])
+							if chkIdx > 0 {
```

Comment:
Ah yea that's much better--done!

Reply:
Depending on the context and how samples are joined together, the counter reset hint for a counter histogram can vary. So yes, it should be good to ignore the hint for counters, but check for gauge (because a gauge histogram always stays as gauge). In case you know exactly or expect how things should be behaving, you may choose to check for the exact counter reset hint as well :)

Reply:
I think in this test we're not testing the reset hint, so ignoring it should be safe and lead to less flaky tests down the line, thanks